### PR TITLE
feat(launch): allow direct configuration of registry uri for all registries

### PR DESCRIPTION
**Pull request by bcsherma**
Fixes
-----
- Fixes WB-14086

Description
-----------
What does the PR do?

### Summary
🧪🛠️🚀

This pull request enhances the functionality and testing of the `ElasticContainerRegistry` and `GoogleArtifactRegistry` classes for launching Docker images from different cloud providers. It adds more ways of specifying the image location, validates the input format, and handles errors gracefully. It also adds `pytest` tests for both classes in `test_ecr.py` and `test_gcp_artifact_registry.py`.

> _We are the masters of the registry_
> _We validate and handle every error_
> _We build our images from any `uri`_
> _We are the `GoogleArtifactRegistry` and the `ElasticContainerRegistry`_

### Walkthrough
*  Add and modify tests for the `ElasticContainerRegistry` and `GoogleArtifactRegistry` classes ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-2c78d986a9982f511cf6944de6e9e034db0df251d85241bd6f763071667a6507L3-R7), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-2c78d986a9982f511cf6944de6e9e034db0df251d85241bd6f763071667a6507R83-R99), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-8cf66776df7de1e2d6967ba612ee06b02d19924fb3312ebfc3bd87e4b44d3df5R37-R68))
  * Test the error handling and validation logic of the `from_config` class methods using `pytest` and `LaunchError` ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-2c78d986a9982f511cf6944de6e9e034db0df251d85241bd6f763071667a6507R83-R99), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-8cf66776df7de1e2d6967ba612ee06b02d19924fb3312ebfc3bd87e4b44d3df5R37-R68))
  * Test the attribute setting and parsing logic of the `from_config` class methods using valid and invalid config dictionaries ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-2c78d986a9982f511cf6944de6e9e034db0df251d85241bd6f763071667a6507R83-R99), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-8cf66776df7de1e2d6967ba612ee06b02d19924fb3312ebfc3bd87e4b44d3df5R37-R68))
*  Update the `from_config` class methods of the `ElasticContainerRegistry` and `GoogleArtifactRegistry` classes in `elastic_container_registry.py` and `google_artifact_registry.py` ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL4-R8), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL55-R58), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL73-R99), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-e48f9a73317248f64c1316ead0a9b85822be710a1a737674e3182c55d510f357R6-R7), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-e48f9a73317248f64c1316ead0a9b85822be710a1a737674e3182c55d510f357L108-R144))
  * Import the `re` and `yaml` modules for parsing and validating the `uri` key and formatting the error messages ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL4-R8), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-e48f9a73317248f64c1316ead0a9b85822be710a1a737674e3182c55d510f357R6-R7))
  * Modify the type annotation of the `config` parameter to specify string keys and values ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL55-R58))
  * Allow the class construction from either a `uri` or a `repository` key, or a `repository` and `image-name` keys, depending on the class ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL73-R99), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-e48f9a73317248f64c1316ead0a9b85822be710a1a737674e3182c55d510f357L108-R144))
  * Validate the format of the `uri` using a regular expression and raise a `LaunchError` if it is invalid ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL73-R99), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-e48f9a73317248f64c1316ead0a9b85822be710a1a737674e3182c55d510f357L108-R144))
  * Raise a `LaunchError` if neither or both keys are provided and provide a more informative error message with the config dump ([link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-0a4454df683c52397f91fcedc1f12f5d1d4ab7d76478361684bbaa893692051cL73-R99), [link](https://github.com/wandb/wandb/pull/5760/files?diff=unified&w=0#diff-e48f9a73317248f64c1316ead0a9b85822be710a1a737674e3182c55d510f357L108-R144))

Testing
-------
How was this PR tested?

Checklist
-------
- [ ] Include reference to internal ticket "Fixes WB-NNNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

### <samp>🤖 Generated by Copilot at 0aafd9f</samp>

> _We are the masters of the registry_
> _We validate and handle every error_
> _We build our images from any `uri`_
> _We are the `GoogleArtifactRegistry` and the `ElasticContainerRegistry`_

### CODE REVIEW

File: wandb/sdk/launch/registry/elastic_container_registry.py [hunk 1]
```diff
@@ -70,10 +73,29 @@ def from_config(  # type: ignore[override]
...
-        if not repository:
+        if ("uri" in config) != ("repository" in config):
+            raise LaunchError(
+                "Could not create ElasticContainerRegistry from config. Either 'uri' or "
+                f"'repository' is required. The config received was:\n{yaml.dump(config)}."
+            )
+        if "repository" in config:
+            repository = config.get("repository")
+        else:
+            match = re.match(
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/registry/elastic_container_registry.py:84]: Same with region

File: wandb/sdk/launch/registry/elastic_container_registry.py [hunk 2]
```diff
@@ -70,10 +73,29 @@ def from_config(  # type: ignore[override]
                 f"Could not create ElasticContainerRegistry from config. Expected type 'ecr' "
                 f"but got '{config.get('type')}'."
             )
-        repository = config.get("repository")
-        if not repository:
+        if ("uri" in config) != ("repository" in config):
+            raise LaunchError(
+                "Could not create ElasticContainerRegistry from config. Either 'uri' or "
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/registry/elastic_container_registry.py:78]: Which also explains why the test is failing

**Review by KyleGoyette — COMMENTED** [review]
If uri is the preferred mechanism we'll also want to update launch-config.yamls to indicate that

File: wandb/sdk/launch/registry/elastic_container_registry.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/registry/elastic_container_registry.py:84]: what we should really do is not force the registry to use the same region configured in the environment
https://wandb.atlassian.net/browse/WB-14166

But yeah for now I will add a check that the region + account match

[the remark this answers is not in the document]
Comment (bcsherma) [reply wandb/sdk/launch/registry/elastic_container_registry.py:84]: donezo

**Review by KyleGoyette — APPROVED** [review]

**Pull request merged**